File: 프로그래머스/2/87946. 피로도/피로도.cpp
#include <string>
#include <vector>

using namespace std;

int best = 0;
vector<bool> used;

void dfs(int k, vector<vector<int>>& dungeons, int cnt) {
    best = max(best, cnt);
    
    for (int i = 0; i < dungeons.size(); i++) {
        if (used[i]) continue;
        if (k < dungeons[i][0]) continue;
        used[i] = true;
        dfs(k - dungeons[i][1], dungeons, cnt+1);
        used[i] = false;
    }
}

int solution(int k, vector<vector<int>> dungeons) {
    used.assign(dungeons.size(), false);
    
    dfs(k, dungeons, 0);
    return best;
}

File: 프로그래머스/2/87946. 피로도/피로도.java
class Solution {
    boolean visited[];
    int answer = -1;
    public int solution(int k, int[][] dungeons) {
        visited = new boolean[dungeons.length];
        dfs(dungeons, k, 0);
        return answer;
    }
    private void dfs(int[][] dungeons, int current, int depth) {
        answer = Math.max(depth, answer);
        
        
        for(int i = 0; i < dungeons.length; i++) {
            if(!visited[i]) {
                if (dungeons[i][0] <= current) {
                    visited[i] = true;
                    dfs(dungeons, current - dungeons[i][1], depth + 1);
                    visited[i] = false;
                }
            }
        }
    }
}